lib/dynamoid/criteria/chain.rb

```diff
@@ -312,16 +312,19 @@ def key_present?
         # Chooses the first GSI found that can be utilized for the query
         # But only do so if projects ALL attributes otherwise we won't
         # get back full data
+        result = false
         source.global_secondary_indexes.each do |_, gsi|
           next unless query.keys.map(&:to_s).include?(gsi.hash_key.to_s) && gsi.projected_attributes == :all
+          next if @range_key.present? && !query_keys.include?(gsi.range_key.to_s)
```

Comment:
Looks like I wanted to check GSI's range key, but not the range key of main table, so it should be

```ruby
next if gsi.range_key.present? &&  ...
```

Does it make sense?

Reply:
If my assumption above is correct there are several issues in proposed solution:

1. Result depends on order of GSI declarations in case when we have GSI with range key and GSI without it and  both of them are suitable (both hash keys are involved into query and range key of the first GSI is involved as well). Only the last GSI in the `source.global_secondary_indexes` list will be returned as result, even if it's GSI without range key
2. If there is only one suitable GSI among all the declared in model (so its range key is involved into query) but it has range key not involved into query - the GSI will be skipped as well what isn't expected result.

Reply:
> Looks like GSI's range key should be checked, but not the range key of the main table. So it should be
> 
> ```ruby
> next if gsi.range_key.present? &&  ...
> ```
> 
> Does it make sense?

That code mean is that it will skip if the range key is already chosen and current GSI's range key is not match query.  So the first time in iteration, this logic behaves like same before modified.
Why do you think GSI's range key should be checked?

Reply:
> 1. Result depends on order of GSI declarations in case when we have GSI with range key and GSI without it and  both of them are suitable (both hash keys are involved into query and range key of the first GSI is involved as well). Only the last GSI in the `source.global_secondary_indexes` list will be returned as result, even if it's GSI without range key

Exactly. But is it troubled?
Please tell me why, for example it makes bad performance etc.

Reply:
> 2\. If there is only one suitable GSI among all the declared in model (so its range key is involved into query) but it has range key not involved into query - the GSI will be skipped as well what isn't expected result.

Is it so? If the assumption is correct, this spec should be failed, isn't it?

https://github.com/Dynamoid/dynamoid/blob/e53f943a39cf62b1f56ac6174e44520e0ce41785/spec/dynamoid/criteria/chain_spec.rb#L641-L646